### ISSUE TITLE
`ldexp` has wrong gradient `0.0`

### ISSUE THREAD

**Issue by Doraehhh** [jax-ml/jax#11467]
`ldexp` has wrong gradient `0.0`. For example,

```py
import jax

def fn(x1):
    x2 = jax.numpy.array(1, dtype=jax.numpy.int32)
    return jax.numpy.ldexp(x1, x2)

x1 = jax.numpy.array(1., dtype=jax.numpy.float32)
print(fn(x1)) # 2.0
print(jax.jacrev(fn)(x1)) # 0.0
```

Because `ldexp` computes `x1 * 2**x2`, it should the gradient `2.0` in this example. It seems that it will return 0 as the gradient for any input.

**Comment by jakevdp** [comment]
Thanks for the report – I think this is probably working as expected, because `ldexp` is a strange function that is not actually well-defined for real numbers, but rather extracts details of the floating-point representation of numbers in a way that is not compatible with autodiff.

If you want a roughly equivalent algebraic function that will be compatible with autodiff, I'd suggest using something like
```python
def almost_ldexp(x1, x2):
  return x1 * 2 ** x2
```

**Comment by Doraehhh** [comment]
Thanks for your explanation!

For this kind of incompatible function with autodiff, I think it'd be better to raise an exception instead of returning all zero gradient. What do you think?

**Comment by jakevdp** [comment]
Sure, that seems like a reasonable option

**Comment by Doraehhh** [comment]
I'm more than happy to help collect such kind of APIs.

**Comment by Doraehhh** [comment]
## Should throw exception


### trim_zeros

`trim_zeros` will return 0 as the gradient. I think it is not compatible (or not implemented?) so it is better to raise an exception

```py
import jax
def fn(input):
    return jax.numpy.trim_zeros(input)
input = jax.numpy.array([1., 2., 3.], dtype=jax.numpy.float32)

res = fn(input)
print(res) # [1. 2. 3.]

grad = jax.jacrev(fn)(input)
print(grad)
# [[0. 0. 0.]
#  [0. 0. 0.]
#  [0. 0. 0.]]
```

### bitcast_convert_type

`bitcast_convert_type` will return 0 as the gradient when the `new_dtype` is the same as `input`. I think this API cannot be differentiated so it may raise an exception.

```py
import jax
def fn(input):
    newtype = jax.numpy.float32
    return jax.lax.bitcast_convert_type(input, newtype)
input = jax.numpy.array([1., 1.], dtype=jax.numpy.float32)

res = fn(input)
print(res) # [1. 1.]

jac = jax.jacrev(fn)(input)
print(jac)
# [[0. 0.]
#  [0. 0.]]
```

### union1d, setdiff1d, setxor1d

I think these three APIs cannot be differentiated, which will return 0 as the gradient for any input

```py
import jax
def fn(ar1):
    ar2 = jax.numpy.array([1, 2, 3])
    return jax.numpy.union1d(ar1, ar2)
ar1 = jax.numpy.array([-2., -1.])

res = fn(ar1)
print(res) # [-2. -1.  1.  2.  3.]

jac = jax.jacrev(fn)(ar1)
print(jac)
# [[0. 0.]
#  [0. 0.]
#  [0. 0.]
#  [0. 0.]
#  [0. 0.]]
```

**Comment by jakevdp** [comment]
Thanks - are you interested in contributing pull requests to fix these issues?

**Comment by YangChenyuan** [comment]
I am interested in fixing these issues. But I am new here, where should I add the exception checker? Are there any material/pull requests that I can refer to? Thanks in advance.

**Comment by jakevdp** [comment]
`trim_zeros` we should be able to fix pretty easily by not concretizing the input

**Comment by carlosgmartin** [comment]
@jakevdp How about a [custom jvp](https://jax.readthedocs.io/en/latest/_autosummary/jax.custom_jvp.html) for [ldexp](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ldexp.html) that simply returns `1 << x2`, i.e. `2 ** x2`? I can submit a PR for that.

**Comment by carlosgmartin** [comment]
Given #23923, I think this can be closed.
